applications/TrilinosApplication/custom_strategies/builder_and_solvers/trilinos_block_builder_and_solver.h

```diff
@@ -4,954 +4,842 @@
 
         KRATOS_CATCH("")
-
     }
 
-    //**************************************************************************
-    //**************************************************************************
-    void BuildLHS(
-        typename TSchemeType::Pointer pScheme,
-        ModelPart& r_model_part,
-        TSystemMatrixType& A) override
+    /**
+     * @brief Function to perform the building of the LHS
+     * @details Depending on the implementation choosen the size of the matrix
+     * could be equal to the total number of Dofs or to the number of
+     * unrestrained dofs
+     * @param pScheme The integration scheme considered
+     * @param rModelPart The model part of the problem to solve
+     * @param rA The LHS matrix
+     */
+    void BuildLHS(typename TSchemeType::Pointer pScheme,
+                  ModelPart& rModelPart,
+                  TSystemMatrixType& rA) override
     {
         KRATOS_TRY
 
-        //getting the elements from the model
-        ElementsArrayType& pElements = r_model_part.Elements();
-
-        //getting the array of the conditions
-        ConditionsArrayType& ConditionsArray = r_model_part.Conditions();
-
-        //resetting to zero the vector of reactions
-// 			TSparseSpace::SetToZero(BaseType::mReactionsVector);
+        const int nelements = static_cast<int>(rModelPart.Elements().size());
+        const int nconditions = static_cast<int>(rModelPart.Conditions().size());
+        // Resetting to zero the vector of reactions
+        TSparseSpace::SetToZero(*BaseType::mpReactionsVector); // TODO: Check if required
 
-        //contributions to the system
-        LocalSystemMatrixType LHS_Contribution = LocalSystemMatrixType(0,0);
-
-        //vector containing the localization in the system of the different
-        //terms
-        Element::EquationIdVectorType EquationId;
+        // Contributions to the system
+        LocalSystemMatrixType LHS_Contribution = LocalSystemMatrixType(0, 0);
+        LocalSystemVectorType RHS_Contribution = LocalSystemVectorType(0);
 
-        ProcessInfo& CurrentProcessInfo = r_model_part.GetProcessInfo();
+        // vector containing the localization in the system of the different terms
+        Element::EquationIdVectorType equation_ids_vector;
+        ProcessInfo& r_current_process_info = rModelPart.GetProcessInfo();
+        ModelPart::ElementsContainerType::iterator el_begin = rModelPart.ElementsBegin();
+        ModelPart::ConditionsContainerType::iterator cond_begin =
+            rModelPart.ConditionsBegin();
 
         // assemble all elements
-        for (typename ElementsArrayType::ptr_iterator it=pElements.ptr_begin(); it!=pElements.ptr_end(); ++it)
-        {
-            //calculate elemental contribution
-            pScheme->Calculate_LHS_Contribution(*it,LHS_Contribution,EquationId,CurrentProcessInfo);
+        for (int k = 0; k < nelements; k++) {
+            ModelPart::ElementsContainerType::iterator it = el_begin + k;
```

Comment:
I would use the ptr iterator, since you are always passing a pointer anyway. You can still make it work with this syntax as
```
ModelPart::ElementsContainerType::ptr_iterator el_begin = rModelPart.Elements().ptr_begin();
auto it = el_begin + k;
```